### PR TITLE
add support for top-p and top-k inference and training

### PR DESCRIPTION
**Pull request by faresobeid**
> [!NOTE]
> Adds configurable top-p/top-k sampling end-to-end (config → orchestrator → batches → trainer) and masks logits accordingly during loss computation; removes min_tokens handling and cleans up sampling args.
> 
> - **Config**:
>   - Add `top_p` and `top_k` to `SamplingConfig`; expose in `EvalSamplingConfig`.
>   - Remove `min_tokens` fields/usages.
> - **Orchestrator**:
>   - Pass `sampling.top_p`/`top_k` into `prepare_batch` and persist in saved micro-batches.
>   - Simplify `get_sampling_args`: always enable `logprobs`, `return_token_ids`, `prompt_logprobs`; set `extra_body['top_k']` from config; drop hardcoded `top_p` and `min_tokens` mapping.
> - **Batch/Data**:
>   - Thread `top_p`/`top_k` through `prepare_micro_batch(_packing)` and `MicroBatch` schema.
> - **Trainer**:
>   - `selective_log_softmax` now accepts `top_p`/`top_k` and applies `_apply_top_k_top_p` masking before log-softmax; training loop supplies these per micro-batch.
> - **Eval/Synthesis**:
>   - Build sampling args with optional `top_k`, `min_p`, `repetition_penalty`; stop setting `min_tokens`.
> - **Tests**:
>   - Update batch test to pass `top_p`/`top_k` and assert padding behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6ecac8f5e2e4860e1ed41bebe54ea735f2acd478. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/orchestrator/batch.py [hunk 1]
```diff
@@ -51,13 +51,15 @@ def prepare_sample(
...
-def prepare_micro_batch(samples: list[MicroBatch], temperature: float):
+def prepare_micro_batch(samples: list[MicroBatch], temperature: float, top_p: float, top_k: int):
     micro_batch = {}
 
     for key in ["input_ids", "advantages", "loss_mask", "inference_logprobs", "position_ids"]:
         micro_batch[key] = torch.stack([sample[key] for sample in samples], dim=0)
 
     micro_batch["temperature"] = temperature
+    micro_batch["top_p"] = top_p
+    micro_batch["top_k"] = top_k
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/batch.py:62]: this gives me so much eye cancer ahaha (not your fault ofc, just this pattern in general). i am really wondering at this point whether we should just have the sampling args as shared config as well, i.e. introduce sampling args on the trainer as well and ensure via rl.py that these are always set to the same values

**Review by mikasenghaas — COMMENTED** [review]
nice! can we do a quick run w/ non-standard values to make sure everything works as expected

File: src/prime_rl/orchestrator/utils.py [hunk 2]
```diff
@@ -31,13 +31,12 @@ def get_sampling_args(sampling_config: SamplingConfig) -> dict:
...
     # https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#extra-parameters_2
     sampling_args = dict(sampling_config)
-    sampling_args["top_p"] = 1.0
     sampling_args["logprobs"] = True
     sampling_args["extra_body"] = {
         **sampling_config.extra_body,
         "return_token_ids": True,  # Always return token IDs
         "prompt_logprobs": True,  # Always return prompt logprobs
-        "top_k": -1,
+        "top_k": sampling_args.pop("top_k"),
```
Comment (samsja) [inline src/prime_rl/orchestrator/utils.py]: should we also add top p here ?

File: src/prime_rl/trainer/rl/loss.py [hunk 3]
```diff
@@ -17,6 +17,61 @@ def selective_log_softmax(
...
+        top_p_mask = probs_sum <= (1.0 - top_p)
+        top_p_mask[:, -1] = False
+        logits_sort = logits_sort.masked_fill(top_p_mask, float("-inf"))
+
+    logits = logits_sort.scatter(dim=-1, index=logits_idx, src=logits_sort)
+    return logits.view(original_shape)
+
+
+@jaxtyped(typechecker=typechecker)
+@torch.compile(dynamic=True)
```
Comment (samsja) [inline src/prime_rl/trainer/rl/loss.py:53]: what is dynamic doing here ?

**Review by samsja — COMMENTED** [review]
pr looks good, added some minor comments, would also want to do a short run with and without top p k and show the mismatch before and after to see if it is good

File: src/prime_rl/orchestrator/utils.py (hunk 2 above)
Comment (faresobeid) [reply src/prime_rl/orchestrator/utils.py]: its already in the sampling args, but tbf idk why top_k is here then. Maybe needed?

File: src/prime_rl/trainer/rl/loss.py (hunk 3 above)
Comment (faresobeid) [reply src/prime_rl/trainer/rl/loss.py:53]: to handle different batch sizes iiuc

**Comment by faresobeid** [comment]
Closed as i realised this doesn't actually work, and for actual support we would need a vllm patch
